sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,426 @@
+static int stat_etag(char *filename, union stattag *stattag);
+
+/* These are almost but but not quite coap_handler_t -- they require the
+ * request to be already fully read and digested into the last argument. */
+
+static ssize_t coapfileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+static ssize_t coapfileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request);
+/** Create a CoAP response for a given errno (eg. EACCESS -> 4.03 Forbidden
+ * etc., defaulting to 5.03 Internal Server Error) */
+static size_t coapfileserver_errno_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, int err);
+/** Return true if path/name is a directory.
+ *
+ * Currently assembling the values in a buffer, this could be done without a
+ * buffer if an openat / vfs_openat was present.
+ * */
+static bool entry_is_dir(char *path, char *name);
+
+static unsigned _count_char(const char *s, char c)
+{
+    unsigned count = 0;
+    while (*s) {
+        if (*s == c) {
+            ++count;
+        }
+        ++s;
+    }
+
+    return count;
+}
+
+ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *ctx) {
+    struct coapfileserver_entry *entry = ctx;
+    struct requestdata request = {
+        .etag_sent = false,
+        .blocknum2 = 0,
+        .szx2 = CONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX,
+    };
+    /** Index in request.namebuf. Must not point at the last entry as that will be
+     * zeroed to get a 0-terminated string. */
+    size_t namelength = 0;
+    bool trailing_slash = false;
+    /** If no path component comes along at all, it'll count as a trailing
+     * slash no matter the trailing_slash value */
+    bool any_component = false;
+    uint8_t errorcode = COAP_CODE_INTERNAL_SERVER_ERROR;
+    uint8_t strip_remaining = _count_char(entry->resource, '/');
+
+    /* If a root directory for the server was specified, use that */
+    if (entry->root && strlen(entry->root)) {
+        strncpy(request.namebuf, entry->root, sizeof(request.namebuf));
+        namelength = strlen(entry->root);
+    }
+
+    coap_optpos_t opt;
+    bool is_first = true;
+    uint8_t *value;
+    ssize_t optlen;
+    while ((optlen = coap_opt_get_next(pdu, &opt, &value, is_first)) !=
+            -ENOENT)
+    {
+        if (optlen < 0) {
+            errorcode = COAP_CODE_BAD_REQUEST;
+            goto error;
+        }
+
+        is_first = false;
+        switch (opt.opt_num) {
+            case COAP_OPT_URI_PATH:
+                if (strip_remaining != 0) {
+                    strip_remaining -= 1;
+                    continue;
+                }
+                if (trailing_slash) {
+                    errorcode = COAP_CODE_BAD_REQUEST;
+                    goto error;
+                }
+                any_component = true;
+                if (optlen == 0) {
+                    trailing_slash = true;
+                    continue;
+                }
```

Comment:
If intermediate slashes are fine, `trailing_slash` and `any_component` can be removed.
```diff
diff --git a/sys/net/application_layer/coapfileserver/coapfileserver.c b/sys/net/application_layer/coapfileserver/coapfileserver.c
index 09f1c3d4de..67735e29d2 100644
--- a/sys/net/application_layer/coapfileserver/coapfileserver.c
+++ b/sys/net/application_layer/coapfileserver/coapfileserver.c
@@ -89,10 +89,8 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
     /** Index in request.namebuf. Must not point at the last entry as that will be
      * zeroed to get a 0-terminated string. */
     size_t namelength = 0;
-    bool trailing_slash = false;
     /** If no path component comes along at all, it'll count as a trailing
      * slash no matter the trailing_slash value */
-    bool any_component = false;
     uint8_t errorcode = COAP_CODE_INTERNAL_SERVER_ERROR;
     uint8_t strip_remaining = _count_char(entry->resource, '/');
 
@@ -104,6 +102,7 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
 
     coap_optpos_t opt;
     bool is_first = true;
+    bool is_directory = true; /* either no path component at all or trailing '/' */
     uint8_t *value;
     ssize_t optlen;
     while ((optlen = coap_opt_get_next(pdu, &opt, &value, is_first)) !=
@@ -121,13 +120,7 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
                     strip_remaining -= 1;
                     continue;
                 }
-                if (trailing_slash) {
-                    errorcode = COAP_CODE_BAD_REQUEST;
-                    goto error;
-                }
-                any_component = true;
-                if (optlen == 0) {
-                    trailing_slash = true;
+                if ((is_directory = (optlen == 0))) { /* '/' */
                     continue;
                 }
                 if (memchr(value, '\0', optlen) != NULL ||
@@ -177,8 +170,6 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
     }
 
     request.namebuf[namelength] = '\0';
-    bool is_directory = trailing_slash || !any_component;
-
     DEBUG("request: '%s'\n", request.namebuf);
 
     /* Note to self: As we parse more options than just Uri-Path, we'll likely
```
`/vfs` --> dirctory `/nvm0`
`/vfs/` --> dirctory `/nvm0`
`/vfs/any` --> file `/nvm0/any`
`/vfs//any` --> file `/nvm0/any`
`/vfs/any/` --> directory `/nvm0/any`
`/vfs//any/` --> directory `/nvm0/any`

Reply:
Failing on inner empty path components is desired (as is failing on slashless access to a dir or vice versa). But unlike the trailing slash error, this could be rare enough as a user error that the optimization you propose is acceptable -- we'd just silently accept erroneous requests.

Reply:
> as is failing on slashless access to a dir or vice versa

Yup, this is at latest catched by `vfs_open()` or `vfs_opendir()`, respectively.

> this could be rare enough as a user error that the optimization you propose is acceptable

Ah ok, for `URI`s it makes a difference but for filesystems it does not. Her I´d say it depends on how you look at it.
We can also keep it like that, I´m fine with either.

Reply:
IMHO the less code the better.

Tbh I didn't fully wrap my head around the whole etag thing and I don't quite see how useful it is, so it's high on my culling list.

Reply:
> Yup, this is at latest catched by vfs_open() or vfs_opendir(), respectively.

I prefer explicitly checking for the trailing slash because the lax handling of file systems is why users start laxly handling this in URIs too, and in URIs that is problematic. I'd be OK with silently ignoring interior empty elements.

Reply:
The trailing slash is still recognized. :)
>  /vfs/any/ --> directory /nvm0/any

With this notation I ment that "/vfs/any/" is handled by `coapfileserver_directory_handler` with `request.namebuf` containig `/vfs/any`.

With anothe simpe `if` we can still handle intermediate slashes.

```diff
diff --git a/sys/net/application_layer/gcoap/fileserver.c b/sys/net/application_layer/gcoap/fileserver.c
index c174413d2e..15d2ce5438 100644
--- a/sys/net/application_layer/gcoap/fileserver.c
+++ b/sys/net/application_layer/gcoap/fileserver.c
@@ -143,7 +143,7 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
         namelength = strlen(entry->root);
     }
 
-    bool is_directory = true; /* either no path component at all or trailing '/' */
+    bool is_directory = false; /* trailing '/' */
     coap_optpos_t opt = {
         .offset = coap_get_total_hdr_len(pdu),
     };
@@ -162,6 +162,10 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
                     strip_remaining -= 1;
                     continue;
                 }
+                if (is_directory) { /* "//" */
+                    errorcode = COAP_CODE_PATH_NOT_FOUND;
+                    goto error;
+                }
                 if ((is_directory = (optlen == 0))) { /* '/' */
                     continue;
                 }
@@ -218,7 +222,7 @@ ssize_t coapfileserver_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, void *
     /* Note to self: As we parse more options than just Uri-Path, we'll likely
      * pass a struct pointer later. So far, those could even be hooked into the
      * resource list, but that'll go away once we parse more options */
-    if (is_directory) {
+    if (is_directory || strcmp(request.namebuf, entry->root ? entry->root : '\0') == 0) {
         return coapfileserver_directory_handler(pdu, buf, len, &request, entry);
     }
     else {
```

```
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs
</vfs/adirectory/>,</vfs/anchor.txt>
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/
</vfs/adirectory/>,</vfs/anchor.txt>
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs//
4.04
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/adirectory
5.00
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/adirectory/
</vfs/adirectory/afile.txt>
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/adirectory//
4.04
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/adirectory/afile.txt
I_am_just_some_data
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/vfs/adirectory//afile.txt
4.04
[fabian@fabianl390yoga ~]$ coap-client -m get -a fe80::3478:9157:9fd5:81c3%eth0 coap://[fe80::4c49:6fff:fe54:0]/
4.04
```